public/app/features/scenes/variables/variants/query/QueryVariable.tsx

```diff
@@ -0,0 +1,184 @@
+      ...initialState,
+    });
+  }
+
+  public activate(): void {
+    super.activate();
+    const timeRange = sceneGraph.getTimeRange(this);
+
+    if (this.state.refresh === VariableRefresh.onTimeRangeChanged) {
+      this._subs.add(
+        timeRange.subscribeToState({
+          next: () => {
+            this.updateSubscription = this.validateAndUpdate().subscribe();
+          },
+        })
+      );
+    }
+  }
+
+  public deactivate(): void {
+    super.deactivate();
+    if (this.updateSubscription) {
+      this.updateSubscription.unsubscribe();
+    }
+
+    if (this.dataSourceSubject) {
+      this.dataSourceSubject.unsubscribe();
+    }
+  }
+
+  public getValueOptions(args: VariableGetOptionsArgs): Observable<VariableValueOption[]> {
+    if (this.state.query === '' || !this.state.datasource) {
+      return of([]);
+    }
+
+    return new Observable<VariableValueOption[]>((observer) => {
+      this.dataSourceSubject = new Subject<DataSourceApi>();
+
+      // Wait for the data source to be ready
+      this.dataSourceSubject.subscribe({
+        next: (ds) => {
+          const runner = createQueryVariableRunner(ds);
+          const target = runner.getTarget(this);
+          const request = this.getRequest(target);
+          runner
+            .runRequest({}, request)
+            .pipe(
+              filter((data) => data.state === LoadingState.Done || data.state === LoadingState.Error), // we only care about done or error for now
+              take(1), // take the first result, using first caused a bug where it in some situations throw an uncaught error because of no results had been received yet
+              mergeMap((data: PanelData) => {
+                if (data.state === LoadingState.Error) {
+                  return throwError(() => data.error);
+                }
+
+                return of(data);
+              }),
+              toMetricFindValues(),
+              mergeMap((values) => {
+                let regex = '';
+                if (this.state.regex) {
+                  regex = sceneGraph.interpolate(this, this.state.regex, undefined, 'regex');
+                }
+                return of(metricNamesToVariableValues(regex, this.state.sort, values));
+              }),
+              catchError((error) => {
+                if (error.cancelled) {
+                  return of([]);
+                }
+
+                return throwError(() => error);
+              })
+            )
+            .subscribe((values) => {
+              observer.next(values);
+              observer.complete();
+            });
+        },
+        error: (e) => observer.error(e),
+      });
+
+      // Resolve variable's data soure
+      getDataSourceSrv()
+        .get(this.state.datasource ?? '')
+        .then((ds) => {
+          this.dataSourceSubject?.next(ds);
+        })
+        .catch((err) => {
+          this.dataSourceSubject?.error(err);
+        });
+    });
```

Comment:
```suggestion
     return from(getDataSourceSrv().get(this.state.datasource ?? '')).pipe(
      mergeMap((ds) => {
        const runner = createQueryVariableRunner(ds);
        const target = runner.getTarget(this);
        const request = this.getRequest(target);

        return runner.runRequest({}, request).pipe(
          filter((data) => data.state === LoadingState.Done || data.state === LoadingState.Error), // we only care about done or error for now
          take(1), // take the first result, using first caused a bug where it in some situations throw an uncaught error because of no results had been received yet
          mergeMap((data: PanelData) => {
            if (data.state === LoadingState.Error) {
              return throwError(() => data.error);
            }

            return of(data);
          }),
          toMetricFindValues(),
          mergeMap((values) => {
            let regex = '';
            if (this.state.regex) {
              regex = sceneGraph.interpolate(this, this.state.regex, undefined, 'regex');
            }
            return of(metricNamesToVariableValues(regex, this.state.sort, values));
          }),
          catchError((error) => {
            if (error.cancelled) {
              return of([]);
            }

            return throwError(() => error);
          })
        );
      })
    );
```